# Welcome
This repository contains all scripts and data tables used for the analysis of viruses in total metagenomes generated from upland peatland soils in the UK, as presented in [Kosmopoulos et al. (2025; in review)]().

All raw sequencing data are publicly available in the NCBI Short Read Archive under BioProject accession [PRJNA1203648](https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA1203648). Whole assembled metagenomic contigs as well as high-quality (CheckM) prokaryotic metagenome-assembled genomes (MAGs) are available at the NCBI WGS using the same BioProject accession. All viral metagenome-assembled genomes (vMAGs) and prokaryotic metagenome-assembled genomes (medium and high quality) are publicly available on figshare under the DOI [10.6084/m9.figshare.28143446](https://doi.org/10.6084/m9.figshare.28143446).

Soil samples, their associated environmental parameters, and metadata were originally obtained and sequenced by [Pallier et al. (2025)](https://doi.org/10.1101/2025.08.01.668219).

# Repository Organization
`Scripts` contains python and shell scripts used to conduct the analyses and run the tools mentioned in the [bioinformatic workflow](https://github.com/AnantharamanLab/UKPeatlandViruses/blob/main/workflow.ipynb).

`Data` contains `.RDS` formatted data that can be loaded into R for analysis. These data include sample metadata as well as virus and host community summary statistics, virus and host genome information, virus and host genome counts, virus-host predictions, gene annotations, read mapping statistics, functional predictions, and more. It also includes the `.shp` shapefiles used to generate the map in Figure 1A.

`Tables` include `.csv` and `.tsv` formatted data tables used in analyses to generate the data files present in `Data`. It also contains organized and summarized data that went into the Supplementary Tables.

`Notebooks` contain R markdown notebooks of all the R code used for the analyses described here. Pre-formatted expected outputs are displayed. A table of contents for these notebooks is given below.

`Plots` contain saved plots (PNG formatted) that were generated by the notebooks above, organized into sub-directories by notebook.

# How to Reproduce the Analyses
All required software packages, their versions, and dependencies are noted in each notebook listed in the table of contents below. For the analyses in the notebook *[Bioinformatic Workflow](https://github.com/AnantharamanLab/UKPeatlandViruses/blob/main/workflow.ipynb)*, software can be installed using [Bioconda](https://bioconda.github.io/). The remaining analyses are performed in R (v4.4.0), and the required packages can be installed from the [Bioconductor](https://www.bioconductor.org/) and [CRAN](https://cran.r-project.org/) repositories. Installing all required packages takes approximately 1â€“2 hours, and significantly less when using [`mamba`](https://github.com/mamba-org/mamba) for Bioconda packages.

To reproduce the analyses in the manuscript, code cells can be copied and pasted into a terminal running a Linux operating system (for commands listed in *[Bioinformatic Workflow](https://github.com/AnantharamanLab/UKPeatlandViruses/blob/main/workflow.ipynb)*) or `.Rmd` files can be opened and run directly in [RStudio](https://posit.co/download/rstudio-desktop/), as the required data and tables are included in this repository.

On a "standard" desktop computer, the R-based analyses should take no more than one hour to run in total. Many commands and software used in *[Bioinformatic Workflow](https://github.com/AnantharamanLab/UKPeatlandViruses/blob/main/workflow.ipynb)* require access to a server with more than 250 GB of RAM and more than 2 TB of disk space.

The total runtime to fully reproduce the analyses (from metagenome assembly, binning, virus identification, and annotations to clustering and read mapping) will depend heavily on the number of available CPU cores, but will generally take at least 3 full days for assembly alone. We recommend obtaining the pre-assembled metagenomes, MAGs, and vMAGs for analysis using the NCBI BioProject and figshare repository listed above.

# Notebook Table of Contents
1. [Bioinformatic workflow](https://github.com/AnantharamanLab/UKPeatlandViruses/blob/main/workflow.ipynb)
2. [Gather and organize data](https://github.com/AnantharamanLab/UKPeatlandViruses/blob/main/Notebooks/data.md)
3. [Soil Environmental Parameters (Extended Data Figure 1)](https://github.com/AnantharamanLab/UKPeatlandViruses/blob/main/Notebooks/environment.md)
4. [Peatland Map and Virus/Host Diversity (Figure 1, Extended Data Figure 2)](https://github.com/AnantharamanLab/UKPeatlandViruses/blob/main/Notebooks/diversity.md)
5. [Virus Distribution (Figure 2)](https://github.com/AnantharamanLab/UKPeatlandViruses/blob/main/Notebooks/virus_distribution.md)
6. [Virus Genome DESeq and Clustering (Extended Data Figure 3)](https://github.com/AnantharamanLab/UKPeatlandViruses/blob/main/Notebooks/virus_deseq.md)
7. [Host Genome DESeq and Clustering (Extended Data Figure 4)](https://github.com/AnantharamanLab/UKPeatlandViruses/blob/main/Notebooks/host_deseq.md)
8. [Virus and Host Relative Abundance (Figure 3)](https://github.com/AnantharamanLab/UKPeatlandViruses/blob/main/Notebooks/virus_host.md)
9. [Virus Genes (Figure 4)](https://github.com/AnantharamanLab/UKPeatlandViruses/blob/main/Notebooks/virus_genes.md)
10. [Virus-Host Abundance Relationships and Lysogenic Viruses (Figure 5, Extended Data Figure 5, Extended Data Figure 6)](https://github.com/AnantharamanLab/UKPeatlandViruses/blob/main/Notebooks/virus_replication.md)

# 
**James C. Kosmopoulos | kosmopoulos \[at\] wisc \[dot\] edu**<br>
**PhD candidate | Microbiology Doctoral Training Program**<br>
**Anantharaman lab | anantharamanlab.wisc.edu**<br>
**Department of Bacteriology | University of Wisconsin - Madison**
